### PR TITLE
add const to parser.d

### PR DESCRIPTION
**Pull request by WalterBright**
trivial

### CODE REVIEW

**Comment by quickfur** [comment]
Does this have performance benefits?

File: src/parse.d [hunk 1]
```diff
@@ -7471,11 +7459,10 @@ public:
 
     Expression parseCmpExp()
     {
-        Expression e;
         Expression e2;
```
Comment (9rnsr) [inline src/parse.d]: D allows to declare variables directly under case statements, without conflict with same name variables in other cases. So you can also declare `e2` in each cases.

File: src/parse.d [hunk 2]
```diff
@@ -6510,15 +6509,13 @@ public:
 
     Expression parseExpression()
     {
-        Expression e;
-        Expression e2;
-        Loc loc = token.loc;
+        auto loc = token.loc;
         //printf("Parser::parseExpression() loc = %d\n", loc.linnum);
-        e = parseAssignExp();
+        Expression e = parseAssignExp();
```
Comment (9rnsr) [inline src/parse.d]: Use `auto`?

File: src/parse.d [hunk 3]
```diff
@@ -6510,15 +6509,13 @@ public:
...
         while (token.value == TOKcomma)
         {
             nextToken();
-            e2 = parseAssignExp();
+            Expression e2 = parseAssignExp();
```
Comment (9rnsr) [inline src/parse.d]: Use `auto`?

**Comment by 9rnsr** [comment]
Other than my comments looks good.

**Comment by WalterBright** [comment]
> Does this have performance benefits?

No. It's a stylistic issue:
1. Should default to declaring variables as const, rather than mutable
2. Should minimize the scope of variables, and not reuse them when declaring a new variable would work - the "single assignment" rule

I'd like to go further with single assignment, but for that D would need a `headconst` storage class.

File: src/parse.d (hunk 1 above)
Comment (WalterBright) [inline src/parse.d]: You're right, I'd forgotten.

**Comment by quickfur** [comment]
What are the downsides to adding `headconst` to the language?

**Comment by WalterBright** [comment]
> What are the downsides

Pros:
1. can do single assignment
2. can interface better with C++

Cons:
1. complexity, bloat, kitchen sink

But also this discussion is off topic for the PR, discussing headconst should go in the n.g.

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Pull request merged**